tools/compare.go

```diff
@@ -15,125 +15,291 @@
+		return fmt.Errorf("parsing range: %w", err)
+	}
+	blockRangeSize, err := blockRange.Size()
+	if err != nil {
+		return fmt.Errorf("checking for valid range: %w", err)
+	}
+	if blockRangeSize == 0 {
+		return fmt.Errorf("invalid block range")
+	}
+
+	// Create stores
+	storeExpected, err := dstore.NewDBinStore(args[0])
+	if err != nil {
+		return fmt.Errorf("unable to create store at path %q: %w", args[0], err)
+	}
+	storeReceived, err := dstore.NewDBinStore(args[1])
+	if err != nil {
+		return fmt.Errorf("unable to create store at path %q: %w", args[1], err)
+	}
+
+	// Walk expected files
+	differentBlocks := make(map[string]block.Range)
+	blocksCountedInChunk := -1
+	chunkIsGood := true
+	rangeNum := 0
+	err = storeExpected.Walk(ctx, "00", func(filename string) (err error) {
+		fileRange, err := strconv.Atoi(filename)
+		if err != nil {
+			return fmt.Errorf("parsing filename: %w", err)
+		}
 
-			for {
-				block, err := blockReader.Read()
-				if err == io.EOF {
-					break
+		// If reached end of range
+		if *blockRange.EndBlock() <= uint64(fileRange) {
+			return dstore.StopIteration
+		}
+
+		// If bundleExpected is valid
+		if blockRange.Contains(uint64(fileRange)) {
+			var wg sync.WaitGroup
+			var bundleErrLock sync.Mutex
+			var bundleReadErr error
+			var expectedBlockHashes []string
+			var expectedBlocks map[string]*pbeth.Block
+			var receivedBlocks map[string]*pbeth.Block
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				expectedBlockHashes, expectedBlocks, err = readBundle(ctx, filename, storeExpected)
+				if err != nil {
+					bundleErrLock.Lock()
+					bundleReadErr = multierr.Append(bundleReadErr, err)
+					bundleErrLock.Unlock()
 				}
+			}()
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				_, receivedBlocks, err = readBundle(ctx, filename, storeReceived)
 				if err != nil {
-					return err
+					bundleErrLock.Lock()
+					bundleReadErr = multierr.Append(bundleReadErr, err)
+					bundleErrLock.Unlock()
 				}
-				blockMap[block.ID()] = block.ToNative().(*pbeth.Block)
+			}()
+			wg.Wait()
+			if bundleReadErr != nil {
+				return fmt.Errorf("reading bundles: %w", bundleReadErr)
 			}
-		}
-		return nil
-	}
 
-	wg := sync.WaitGroup{}
-	wg.Add(2)
-	go func() {
-		defer wg.Done()
-		_ = collectBlocks(storeA, blocksA)
-	}()
+			// check if all blocks exists
+			bundleHasDiff := false
 
-	go func() {
-		defer wg.Done()
-		_ = collectBlocks(storeB, blocksB)
-	}()
-	wg.Wait()
+			for _, expectedBlockHash := range expectedBlockHashes {
+				blocksCountedInChunk++
+				expectedBlock := expectedBlocks[expectedBlockHash]
+				receivedBlock, existsInReceived := receivedBlocks[expectedBlockHash]
 
-	for blockID, blockA := range blocksA {
-		blockB, exists := blocksB[blockID]
-		if !exists {
-			continue
-		}
+				// Reset chunk, print if good
+				if blocksCountedInChunk >= 100000 || uint64(blocksCountedInChunk) >= (blockRangeSize-uint64(rangeNum*100000)-1) || uint64(blocksCountedInChunk) == blockRangeSize-1 {
```

Comment:
You should try to avoid "magic numbers", they are hard to reason about because there is no semantic name attached to the value, moreover, you are duplicating the actual value at many places, so it makes changing the constant to something else harder.